lib/matplotlib/backend_tools.py

```diff
@@ -0,0 +1,882 @@
+"""
+Abstract base classes define the primitives for Tools.
+These tools are used by `NavigationBase`
+
+:class:`ToolBase`
+    Simple stateless tool
+
+:class:`ToolToggleBase`
+    Tool that has two states, only one Toggle tool can be
+    active at any given time for the same `Navigation`
+"""
+
+
+from matplotlib import rcParams
+from matplotlib._pylab_helpers import Gcf
+import matplotlib.cbook as cbook
+from weakref import WeakKeyDictionary
+import numpy as np
+
+
+class Cursors:
+    """Simple namespace for cursor reference"""
+    HAND, POINTER, SELECT_REGION, MOVE = list(range(4))
+cursors = Cursors()
+
+
+class ToolBase(object):
+    """Base tool class
+
+    A base tool, only implements `trigger` method or not method at all.
+    The tool is instantiated by `matplotlib.backend_bases.NavigationBase`
+
+    Attributes
+    ----------
+    navigation: `matplotlib.backend_bases.NavigationBase`
+        Navigation that controls this Tool
+    figure: `FigureCanvas`
+        Figure instance that is affected by this Tool
+    name: String
+        Used as **Id** of the tool, has to be unique among tools of the same
+        Navigation
+    """
+
+    keymap = None
+    """Keymap to associate with this tool
+
+    **String**: List of comma separated keys that will be used to call this
+    tool when the keypress event of *self.figure.canvas* is emited
+    """
+
+    description = None
+    """Description of the Tool
+
+    **String**: If the Tool is included in the Toolbar this text is used
+    as a Tooltip
+    """
+
+    image = None
+    """Filename of the image
+
+    **String**: Filename of the image to use in the toolbar. If None, the
+    `name` is used as a label in the toolbar button
+    """
+
+    def __init__(self, navigation, name):
+        self._name = name
+        self.figure = None
+        self.navigation = navigation
+        self.set_figure(navigation.canvas.figure)
+
+    def trigger(self, sender, event, data=None):
+        """Called when this tool gets used
+
+        This method is called by
+        `matplotlib.backend_bases.NavigationBase.tool_trigger_event`
+
+        Parameters
+        ----------
+        event : `Event`
+            The Canvas event that caused this tool to be called
+        sender: object
+            Object that requested the tool to be triggered
+        data: object
+            Extra data
+        """
+
+        pass
+
+    def set_figure(self, figure):
```

Comment:
@fariza Okay, I will take your word for that, just when you say "``set_figure`` is one of those methods that are prone to be overriden for simple tool creation", yet I don't see any of that here, why this would become a common thing for users to want to do.  If this part of the API get left out until MEP23?  Better to later expand the API than to change it.